tests/FSharp.Compiler.ComponentTests/Language/ModuleAsTypeTests.fs

```diff
@@ -0,0 +1,87 @@
+﻿module FSharp.Compiler.ComponentTests.Language.ModuleAsTypeTests
+
+open Xunit
+open FSharp.Test.Compiler
+
+[<Theory>]
+[<InlineData("typeof")>]
+[<InlineData("typedefof")>]
+let ``Module can be used in typeof in lang preview`` (typeofKeyword: string) =
+    Fsx $"""
+let actual = {typeofKeyword}<module FSharp.Core.LanguagePrimitives>.Name
+let expected = "LanguagePrimitives"
+if actual <> expected then failwith $"Expected '{{expected}}', but got '{{actual}}'"
+    """
+    |> asExe
+    |> withLangVersionPreview
+    |> compileAndRun
+    |> shouldSucceed
+
+[<Fact>]
+let ``Module cannot be used in typeof in lang version70`` () =
+    Fsx """
+let actual = typeof<module FSharp.Core.LanguagePrimitives>
+    """
+    |> withLangVersion70
+    |> typecheck
+    |> shouldFail
+    |> withDiagnostics [
+        (Error 3350, Line 2, Col 21, Line 2, Col 27, "Feature 'Allow modules to be used in type application' is not available in F# 7.0. Please use language version 'PREVIEW' or greater.")
+    ]
+
+[<Fact>]
+let ``Module cannot be used in type annotation in lang preview`` () =
+    Fsx """
+let actual: module LanguagePrimitives = Unchecked.defaultof<_>
+    """
+    |> withLangVersionPreview
+    |> typecheck
+    |> shouldFail
+    |> withDiagnostics [
+        (Error 3561, Line 2, Col 13, Line 2, Col 38, "Module references are intended to be used in 'typeof'.")
+    ]
+
+[<Fact>]
+let ``Module can be used as generic type argument in lang preview`` () =
+    Fsx """
+let actual = ResizeArray<module LanguagePrimitives>().GetType().FullName
```

Comment:
Is it expected to be correct? What does this code generate?
I think we should only allow it in `typeof` and `typedefof`? We could check that during type check and produce an error.

Reply:
Well, modules are classes, so the IL is valid. While C# doesn't allow static types as type arguments, F# doesn't mind.

Reply:
And if we decide `typeof` is the only valid use, why not just introduce `moduleof` and avoid mucking with all these parsing changes.